java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToGCS.java

```diff
@@ -0,0 +1,145 @@
+
+  public KafkaToGCS() {
+
+    gcsOutputLocation = getProperties().getProperty(KAFKA_GCS_OUTPUT_LOCATION);
+    gcsOutputFormat = getProperties().getProperty(KAFKA_GCS_OUTPUT_FORMAT);
+    kafkaBootstrapServers = getProperties().getProperty(KAFKA_BOOTSTRAP_SERVERS);
+    kafkaTopic = getProperties().getProperty(KAFKA_TOPIC);
+    kafkaMessageFormat = getProperties().getProperty(KAFKA_MESSAGE_FORMAT);
+    kafkaSchemaUrl = getProperties().getProperty(KAFKA_SCHEMA_URL);
+    gcsCheckpointLocation = gcsOutputLocation.concat("/checkpoint/");
+    kafkaStartingOffsets = getProperties().getProperty(KAFKA_STARTING_OFFSET);
+    kafkaOutputMode = getProperties().getProperty(KAFKA_GCS_OUTPUT_MODE);
+    kafkaAwaitTerminationTimeout =
+        Long.valueOf(getProperties().getProperty(KAFKA_GCS_AWAIT_TERMINATION_TIMEOUT));
+  }
+
+  @Override
+  public void runTemplate() {
+    if (StringUtils.isAllBlank(gcsOutputLocation)
+        || StringUtils.isAllBlank(kafkaBootstrapServers)
+        || StringUtils.isAllBlank(kafkaTopic)) {
+      LOGGER.error(
+          "{},{},{} is required parameter. ",
+          KAFKA_GCS_OUTPUT_LOCATION,
+          KAFKA_BOOTSTRAP_SERVERS,
+          KAFKA_TOPIC);
+      throw new IllegalArgumentException(
+          "Required parameters for KafkaToGCS not passed. "
+              + "Set mandatory parameter for KafkaToGCS template "
+              + "in resources/conf/template.properties file.");
+    }
+
+    if (kafkaMessageFormat.equals("json") & StringUtils.isAllBlank(kafkaSchemaUrl)) {
+      LOGGER.error("{} is a required parameter for JSON format messages", KAFKA_SCHEMA_URL);
+      throw new IllegalArgumentException("Required parameters for KafkaToGCS not passed.");
+    }
+
+    SparkSession spark = null;
+    LOGGER.info(
+        "Starting Kafka to GCS spark job with following parameters:"
+            + "1. {}:{}"
+            + "2. {}:{}"
+            + "3. {}:{}"
+            + "4. {},{}"
+            + "5. {},{}"
+            + "6. {},{}"
+            + "7, {},{}",
+        KAFKA_GCS_OUTPUT_LOCATION,
+        gcsOutputLocation,
+        KAFKA_GCS_OUTPUT_FORMAT,
+        gcsOutputFormat,
+        KAFKA_BOOTSTRAP_SERVERS,
+        kafkaBootstrapServers,
+        KAFKA_TOPIC,
+        kafkaTopic,
+        KAFKA_STARTING_OFFSET,
+        kafkaStartingOffsets,
+        KAFKA_GCS_OUTPUT_MODE,
+        kafkaOutputMode,
+        KAFKA_GCS_AWAIT_TERMINATION_TIMEOUT,
+        kafkaAwaitTerminationTimeout);
+
+    try {
+      // Initialize Spark session
+      spark = SparkSession.builder().appName("Spark KafkaToGCS Job").getOrCreate();
+
+      LOGGER.debug("added jars : {}", spark.sparkContext().addedJars().keys());
+
+      KafkaReader reader = new KafkaReader();
+
+      LOGGER.info("Calling Kafka Reader");
+
+      // Reading Kafka stream into dataframe
+      Dataset<Row> processedData = reader.readKafkaTopic(spark, getProperties());
+
+      // Write the output to GCS location
+      processedData
```

Comment:
Question:
will this code allow multiple threads (executors) to run in parallel ?
If so will each executor write independent file in gcs ?

Reply:
Spark handles spawning additional parallel executors to scale for multiple partitions or data volume